File: app/src/main/java/com/example/englishhindi/database/AchievementRepository.java
package com.bhashasetu.app.database;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import com.bhashasetu.app.model.Achievement;
import com.bhashasetu.app.model.AchievementRegistry;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.List;

/**
 * Repository for managing Achievement data.
 */
public class AchievementRepository {

    private AchievementDao achievementDao;
    private LiveData<List<Achievement>> allAchievements;
    private LiveData<List<Achievement>> unlockedAchievements;
    private LiveData<Integer> totalPoints;
    private LiveData<Integer> unlockedCount;
    private LiveData<Integer> totalCount;
    private ExecutorService executor;

    /**
     * Constructor.
     * 
     * @param application The application context
     */
    public AchievementRepository(Application application) {
        AppDatabase database = AppDatabase.getInstance(application);
        achievementDao = database.achievementDao();
        
        allAchievements = achievementDao.getAllAchievements();
        unlockedAchievements = achievementDao.getUnlockedAchievements();
        totalPoints = achievementDao.getTotalPoints();
        unlockedCount = achievementDao.getUnlockedCount();
        totalCount = achievementDao.getTotalCount();
        
        executor = Executors.newSingleThreadExecutor();
        
        // Initialize achievements if needed
        executor.execute(new InitializeAchievementsTask());
    }

    /**
     * Get all achievements.
     * 
     * @return LiveData list of all achievements
     */
    public LiveData<List<Achievement>> getAllAchievements() {
        return allAchievements;
    }

    /**
     * Get all unlocked achievements.
     * 
     * @return LiveData list of unlocked achievements
     */
    public LiveData<List<Achievement>> getUnlockedAchievements() {
        return unlockedAchievements;
    }

    /**
     * Get achievements by type.
     * 
     * @param type The achievement type
     * @return LiveData list of achievements of the specified type
     */
    public LiveData<List<Achievement>> getAchievementsByType(String type) {
        return achievementDao.getAchievementsByType(type);
    }

    /**
     * Get an achievement by ID.
     * 
     * @param id The achievement ID
     * @return LiveData with the achievement
     */
    public LiveData<Achievement> getAchievementById(String id) {
        return achievementDao.getAchievementById(id);
    }

    /**
     * Get the total achievement points.
     * 
     * @return LiveData with the total points
     */
    public LiveData<Integer> getTotalPoints() {
        return totalPoints;
    }

    /**
     * Get the count of unlocked achievements.
     * 
     * @return LiveData with the unlocked count
     */
    public LiveData<Integer> getUnlockedCount() {
        return unlockedCount;
    }

    /**
     * Get the total count of achievements.
     * 
     * @return LiveData with the total count
     */
    public LiveData<Integer> getTotalCount() {
        return totalCount;
    }

    /**
     * Update an achievement's progress.
     * 
     * @param achievementId The achievement ID
     * @param increment The amount to increment progress by
     * @param callback Callback to be notified if the achievement was unlocked
     */
    public void updateAchievementProgress(String achievementId, int increment, 
                                         OnAchievementUnlockedListener callback) {
        executor.execute(() -> {
            Achievement achievement = achievementDao.getAchievementByIdSync(achievementId);
            if (achievement != null) {
                boolean wasUnlocked = achievement.isUnlocked();
                achievement.incrementProgress(increment);
                achievementDao.update(achievement);
                
                // Notify callback if the achievement was just unlocked
                if (!wasUnlocked && achievement.isUnlocked() && callback != null) {
                    callback.onAchievementUnlocked(achievement);
                }
            }
        });
    }
    
    /**
     * Check and update progress for all achievements of a specific type.
     * 
     * @param type The achievement type
     * @param progressChecker Function to check if an achievement's progress should be updated
     * @param callback Callback to be notified when an achievement is unlocked
     */
    public void checkTypeAchievements(String type, int progressToAdd, 
                                     OnAchievementUnlockedListener callback) {
        executor.execute(() -> {
            List<Achievement> achievements = achievementDao.getAchievementsByType(type).getValue();
            if (achievements != null) {
                for (Achievement achievement : achievements) {
                    if (!achievement.isUnlocked()) {
                        boolean wasUnlocked = achievement.isUnlocked();
                        achievement.incrementProgress(progressToAdd);
                        achievementDao.update(achievement);
                        
                        // Notify callback if the achievement was just unlocked
                        if (!wasUnlocked && achievement.isUnlocked() && callback != null) {
                            callback.onAchievementUnlocked(achievement);
                        }
                    }
                }
            }
        });
    }

    /**
     * Unlock an achievement.
     * 
     * @param achievementId The achievement ID
     * @param callback Callback to be notified when the achievement is unlocked
     */
    public void unlockAchievement(String achievementId, OnAchievementUnlockedListener callback) {
        executor.execute(() -> {
            Achievement achievement = achievementDao.getAchievementByIdSync(achievementId);
            if (achievement != null && !achievement.isUnlocked()) {
                achievement.setUnlocked(true);
                achievement.setProgressCurrent(achievement.getProgressTarget());
                achievementDao.update(achievement);
                
                // Notify callback
                if (callback != null) {
                    callback.onAchievementUnlocked(achievement);
                }
            }
        });
    }

    /**
     * Initialize achievements in the database.
     */
    private class InitializeAchievementsTask implements Runnable {
        @Override
        public void run() {
            // Check if we need to initialize
            if (achievementDao.getTotalCount().getValue() == null || 
                achievementDao.getTotalCount().getValue() == 0) {
                
                // Clear existing achievements (if any)
                achievementDao.deleteAll();
                
                // Insert all standard achievements
                List<Achievement> achievements = AchievementRegistry.createAllAchievements();
                achievementDao.insertAll(achievements);
            }
        }
    }

    /**
     * Listener for achievement unlocked events.
     */
    public interface OnAchievementUnlockedListener {
        void onAchievementUnlocked(Achievement achievement);
    }
}

File: app/src/main/java/com/bhashasetu/app/data/repository/AchievementRepository.kt
package com.bhashasetu.app.data.repository

import android.app.Application
import androidx.lifecycle.LiveData
import com.bhashasetu.app.database.AppDatabase
import com.bhashasetu.app.database.AchievementDao as LegacyAchievementDao
import com.bhashasetu.app.database.ModernAchievementDao
import com.bhashasetu.app.model.Achievement as LegacyAchievement
import com.bhashasetu.app.data.model.Achievement as ModernAchievement
import kotlinx.coroutines.flow.Flow
import jakarta.inject.Singleton
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map

/**
 * Repository for managing both Legacy and Modern Achievement systems.
 *
 * DUAL ACHIEVEMENT SYSTEM:
 * - Legacy System: Uses legacy_achievements table for backward compatibility
 * - Modern System: Uses achievements table for new bilingual features
 *
 * USAGE GUIDELINES:
 * - Use Legacy methods for existing achievement functionality
 * - Use Modern methods for new features (bilingual, categories, analytics)
 * - Use Unified methods when you need to work with both systems together
 */
@Singleton
class AchievementRepository {

    private val legacyDao: LegacyAchievementDao
    private val modernDao: ModernAchievementDao

    // ===============================
    // CONSTRUCTORS
    // ===============================

    /**
     * Primary constructor - takes Application context
     */
    constructor(application: Application) {
        val database = AppDatabase.getInstance(application)
        legacyDao = database.achievementDao()
        modernDao = database.modernAchievementDao()
    }

    /**
     * Alternative constructor - takes both DAOs directly (for testing)
     */
    constructor(legacyDao: LegacyAchievementDao, modernDao: ModernAchievementDao) {
        this.legacyDao = legacyDao
        this.modernDao = modernDao
    }

    // ===============================
    // LEGACY ACHIEVEMENT SYSTEM (Backward Compatibility)
    // ===============================

    /**
     * Get all legacy achievements
     */
    val allLegacyAchievements: LiveData<List<LegacyAchievement>>
        get() = legacyDao.getAllAchievements()

    /**
     * Get unlocked legacy achievements
     */
    val unlockedLegacyAchievements: LiveData<List<LegacyAchievement>>
        get() = legacyDao.getUnlockedAchievements()

    /**
     * Get locked legacy achievements
     */
    val lockedLegacyAchievements: LiveData<List<LegacyAchievement>>
        get() = legacyDao.getAchievementsByUnlockStatus(false)

    /**
     * Get count of unlocked legacy achievements
     */
    val unlockedLegacyAchievementCount: LiveData<Int>
        get() = legacyDao.getUnlockedCount()

    /**
     * Get total points from legacy achievements
     */
    val totalLegacyPoints: LiveData<Int>
        get() = legacyDao.getTotalPoints()

    // Legacy Achievement Operations
    suspend fun insertLegacyAchievement(achievement: LegacyAchievement) {
        legacyDao.insert(achievement)
    }

    suspend fun updateLegacyAchievement(achievement: LegacyAchievement) {
        legacyDao.update(achievement)
    }

    suspend fun deleteLegacyAchievement(achievement: LegacyAchievement) {
        legacyDao.delete(achievement)
    }

    fun getLegacyAchievementById(id: Long): LiveData<LegacyAchievement> {
        return legacyDao.getAchievementById(id)
    }

    suspend fun unlockLegacyAchievement(achievementId: Long, timestamp: Long = System.currentTimeMillis()) {
        legacyDao.unlockAchievement(achievementId, timestamp)
    }

    suspend fun updateLegacyAchievementProgress(achievementId: Long, progress: Int) {
        legacyDao.updateProgress(achievementId, progress)
    }

    fun getLegacyAchievementsByType(type: String): LiveData<List<LegacyAchievement>> {
        return legacyDao.getAchievementsByType(type)
    }

    // ===============================
    // MODERN ACHIEVEMENT SYSTEM (New Features)
    // ===============================

    /**
     * Get all modern achievements
     */
    val allModernAchievements: LiveData<List<ModernAchievement>>
        get() = modernDao.getAllAchievements()

    /**
     * Get unlocked modern achievements
     */
    val unlockedModernAchievements: LiveData<List<ModernAchievement>>
        get() = modernDao.getUnlockedAchievements()

    /**
     * Get locked visible modern achievements
     */
    val lockedModernAchievements: LiveData<List<ModernAchievement>>
        get() = modernDao.getLockedVisibleAchievements()

    /**
     * Get all visible modern achievements (not hidden)
     */
    val visibleModernAchievements: LiveData<List<ModernAchievement>>
        get() = modernDao.getVisibleAchievements()

    /**
     * Get count of unlocked modern achievements
     */
    val unlockedModernAchievementCount: LiveData<Int>
        get() = modernDao.getUnlockedCount()

    /**
     * Get total points from modern achievements
     */
    val totalModernPoints: LiveData<Int>
        get() = modernDao.getTotalPointsFromAchievements()

    /**
     * Get completion percentage for modern achievements
     */
    val modernCompletionPercentage: LiveData<Double>
        get() = modernDao.getCompletionPercentage()

    /**
     * Get all categories from modern achievements
     */
    val allModernCategories: LiveData<List<String>>
        get() = modernDao.getAllCategories()

    // Modern Achievement Operations
    suspend fun insertModernAchievement(achievement: ModernAchievement) {
        modernDao.insert(achievement)
    }

    suspend fun updateModernAchievement(achievement: ModernAchievement) {
        modernDao.update(achievement)
    }

    suspend fun deleteModernAchievement(achievement: ModernAchievement) {
        modernDao.delete(achievement)
    }

    fun getModernAchievementById(id: String): LiveData<ModernAchievement> {
        return modernDao.getAchievementById(id)
    }

    suspend fun unlockModernAchievement(achievementId: String, timestamp: Long = System.currentTimeMillis()) {
        modernDao.unlockAchievement(achievementId, timestamp)
    }

    suspend fun incrementModernAchievementProgress(achievementId: String, increment: Int) {
        modernDao.incrementProgress(achievementId, increment)
    }

    suspend fun setModernAchievementProgress(achievementId: String, progress: Int) {
        modernDao.setProgress(achievementId, progress)
    }

    fun getModernAchievementsByCategory(category: String): LiveData<List<ModernAchievement>> {
        return modernDao.getAchievementsByCategory(category)
    }

    fun getModernAchievementsByType(type: String): LiveData<List<ModernAchievement>> {
        return modernDao.getAchievementsByType(type)
    }

    fun searchModernAchievements(query: String): LiveData<List<ModernAchievement>> {
        return modernDao.searchAchievements(query)
    }

    fun getModernAchievementsNearCompletion(percentage: Double = 80.0): LiveData<List<ModernAchievement>> {
        return modernDao.getAchievementsNearCompletion(percentage)
    }

    suspend fun getModernAchievementsReadyToUnlock(): List<ModernAchievement> {
        return modernDao.getAchievementsReadyToUnlock()
    }

    // ===============================
    // UNIFIED METHODS (Work with Both Systems)
    // ===============================

    /**
     * Get total points from both achievement systems combined
     */
    fun getTotalCombinedPoints(): LiveData<Int> {
        return combine(
            totalLegacyPoints,
            totalModernPoints
        ) { legacy, modern ->
            (legacy ?: 0) + (modern ?: 0)
        }
    }

    /**
     * Get total unlocked achievements from both systems
     */
    fun getTotalUnlockedCount(): LiveData<Int> {
        return combine(
            unlockedLegacyAchievementCount,
            unlockedModernAchievementCount
        ) { legacy, modern ->
            (legacy ?: 0) + (modern ?: 0)
        }
    }

    /**
     * Check if user has any achievements unlocked (either system)
     */
    fun hasAnyAchievements(): LiveData<Boolean> {
        return getTotalUnlockedCount().map { count -> count > 0 }
    }

    /**
     * Get recent achievements from both systems (combined)
     */
    fun getRecentAchievements(limit: Int = 5): LiveData<RecentAchievementsResult> {
        return combine(
            legacyDao.getRecentlyUnlocked(System.currentTimeMillis() - 86400000L, limit), // Last 24 hours
            modernDao.getRecentUnlockedWithMessages(limit)
        ) { legacy, modern ->
            RecentAchievementsResult(
                legacyAchievements = legacy,
                modernAchievements = modern
            )
        }
    }

    // ===============================
    // MAINTENANCE AND UTILITY METHODS
    // ===============================

    /**
     * Clear all achievement data (both systems) - for testing only
     */
    suspend fun clearAllAchievements() {
        legacyDao.deleteAll()
        modernDao.deleteAll()
    }

    /**
     * Reset all achievement progress (both systems) - for testing only
     */
    suspend fun resetAllProgress() {
        modernDao.resetAllProgress()
        // Legacy system doesn't have reset method, would need to update each individually
    }

    // ===============================
    // BACKWARD COMPATIBILITY METHODS
    // ===============================

    /**
     * Maintain backward compatibility with your existing code
     * These methods work with the modern system but keep the same interface
     */

    @Deprecated("Use allModernAchievements for new code", ReplaceWith("allModernAchievements"))
    val allAchievements: LiveData<List<ModernAchievement>>
        get() = allModernAchievements

    @Deprecated("Use unlockedModernAchievements for new code", ReplaceWith("unlockedModernAchievements"))
    val unlockedAchievements: LiveData<List<ModernAchievement>>
        get() = unlockedModernAchievements

    @Deprecated("Use lockedModernAchievements for new code", ReplaceWith("lockedModernAchievements"))
    val lockedAchievements: LiveData<List<ModernAchievement>>
        get() = lockedModernAchievements

    @Deprecated("Use unlockedModernAchievementCount for new code", ReplaceWith("unlockedModernAchievementCount"))
    val unlockedAchievementCount: LiveData<Int>
        get() = unlockedModernAchievementCount

    @Deprecated("Use insertModernAchievement for new code", ReplaceWith("insertModernAchievement(achievement)"))
    suspend fun insert(achievement: ModernAchievement) {
        insertModernAchievement(achievement)
    }

    @Deprecated("Use updateModernAchievement for new code", ReplaceWith("updateModernAchievement(achievement)"))
    suspend fun update(achievement: ModernAchievement) {
        updateModernAchievement(achievement)
    }

    @Deprecated("Use deleteModernAchievement for new code", ReplaceWith("deleteModernAchievement(achievement)"))
    suspend fun delete(achievement: ModernAchievement) {
        deleteModernAchievement(achievement)
    }

    // Note: This method signature is different due to ID type change (Int vs String)
    // You'll need to update calling code to use String IDs for modern achievements
    fun getAchievementById(id: String): LiveData<ModernAchievement> {
        return getModernAchievementById(id)
    }

    // Note: This method signature is different due to ID type change
    // You'll need to update calling code to use String IDs for modern achievements
    suspend fun unlockAchievement(achievementId: String, timestamp: Long = System.currentTimeMillis()) {
        unlockModernAchievement(achievementId, timestamp)
    }

    // ===============================
    // DATA CLASSES
    // ===============================

    /**
     * Result class for combined recent achievements
     */
    data class RecentAchievementsResult(
        val legacyAchievements: List<LegacyAchievement>,
        val modernAchievements: List<ModernAchievement>
    )
}